File: src/prod/src/ServiceModel/reliability/failover/ServiceCorrelationDescription.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace std;
using namespace Common;
using namespace Reliability;

INITIALIZE_SIZE_ESTIMATION(ServiceCorrelationDescription)

ServiceCorrelationDescription::ServiceCorrelationDescription()
    : serviceName_(), 
    scheme_(FABRIC_SERVICE_CORRELATION_SCHEME_AFFINITY)
{
}

ServiceCorrelationDescription::ServiceCorrelationDescription(
    std::wstring const& serviceName, 
    FABRIC_SERVICE_CORRELATION_SCHEME scheme)
    : serviceName_(serviceName), 
    scheme_(scheme)
{
}

ServiceCorrelationDescription::ServiceCorrelationDescription(ServiceCorrelationDescription const& other)
    : serviceName_(other.serviceName_), 
    scheme_(other.scheme_)
{
}

ServiceCorrelationDescription::ServiceCorrelationDescription(ServiceCorrelationDescription && other)
    : serviceName_(std::move(other.serviceName_)), 
    scheme_(other.scheme_)
{
}

ServiceCorrelationDescription & ServiceCorrelationDescription::operator = (ServiceCorrelationDescription const& other)
{
    if (this != &other)
    {
        serviceName_ = other.serviceName_;
        scheme_ = other.scheme_;
    }

    return *this;
}

ServiceCorrelationDescription & ServiceCorrelationDescription::operator = (ServiceCorrelationDescription && other)
{
    if (this != &other)
    {
        serviceName_ = move(other.serviceName_);
        scheme_ = other.scheme_;
    }

    return *this;
}

bool ServiceCorrelationDescription::operator == (ServiceCorrelationDescription const & other) const
{

    if (!StringUtility::AreEqualCaseInsensitive(this->ServiceName, other.ServiceName))
    {
        return false;
    }

    if (this->Scheme != other.Scheme)
    {
        return false;
    }

    return true;
}

void ServiceCorrelationDescription::WriteToEtw(uint16 contextSequenceId) const
{
    ServiceModel::ServiceModelEventSource::Trace->ServiceCorrelationDescription(
        contextSequenceId,
        serviceName_,
        static_cast<int>(scheme_));
}

void ServiceCorrelationDescription::WriteTo(TextWriter& writer, FormatOptions const&) const
{
    writer.Write("{0} {1}", serviceName_, static_cast<uint>(scheme_));
}

File: src/prod/src/managed/Api/src/System/Fabric/Description/ServiceCorrelationDescription.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Description
{
    using System;
    using System.Fabric.Common;
    using System.Fabric.Interop;
    using System.Fabric.Strings;
    using System.Globalization;
    using System.Text;

    /// <summary>
    /// <para>Creates a particular correlation between services.</para>
    /// </summary>
    public sealed class ServiceCorrelationDescription
    {
        private Uri serviceName;
        private ServiceCorrelationScheme scheme;

        /// <summary>
        /// <para>Initializes a new instance of the <see cref="System.Fabric.Description.ServiceCorrelationDescription" /> class.</para>
        /// </summary>
        public ServiceCorrelationDescription()
        {
            this.serviceName = null;
            this.scheme = ServiceCorrelationScheme.Invalid;
        }

        internal ServiceCorrelationDescription(ServiceCorrelationDescription other)
        {
            this.ServiceName = other.ServiceName;
            this.Scheme = other.Scheme;
        }

        /// <summary>
        /// <para>Gets or sets the name of the service that you want to establish the correlation relationship with.</para>
        /// </summary>
        /// <value>
        /// <para>The name of the service that you want to establish the correlation relationship with.</para>
        /// </value>
        /// <remarks>
        ///   <para />
        /// </remarks>
        public Uri ServiceName
        {
            get
            {
                return this.serviceName;
            }

            set
            {
                this.serviceName = value;
            }
        }

        /// <summary>
        /// <para>Gets or sets the <see cref="System.Fabric.Description.ServiceCorrelationScheme" /> which describes the relationship between this 
        /// service and the service specified via <see cref="System.Fabric.Description.ServiceCorrelationDescription.ServiceName" />.</para>
        /// </summary>
        /// <value>
        /// <para>The service correlation scheme.</para>
        /// </value>
        /// <remarks>
        ///   <para />
        /// </remarks>
        public ServiceCorrelationScheme Scheme
        {
            get
            {
                return this.scheme;
            }

            set
            {
                this.scheme = value;
            }
        }

        internal static void Validate(ServiceCorrelationDescription serviceCorrelationDescription)
        {
            Requires.Argument<ServiceCorrelationDescription>("serviceCorrelationDescription", serviceCorrelationDescription).NotNull();
            Requires.Argument<Uri>("serviceCorrelationDescription.Name", serviceCorrelationDescription.ServiceName).NotNullOrWhiteSpace();
        }

        internal static unsafe ServiceCorrelationDescription CreateFromNative(IntPtr nativePtr)
        {
            ReleaseAssert.AssertIfNot(nativePtr != IntPtr.Zero, StringResources.Error_NullNativePointer);

            ServiceCorrelationDescription correlation = new ServiceCorrelationDescription();
            NativeTypes.FABRIC_SERVICE_CORRELATION_DESCRIPTION* casted = (NativeTypes.FABRIC_SERVICE_CORRELATION_DESCRIPTION*)nativePtr;

            correlation.ServiceName = NativeTypes.FromNativeUri(casted->ServiceName);

            switch (casted->Scheme)
            {
                case NativeTypes.FABRIC_SERVICE_CORRELATION_SCHEME.FABRIC_SERVICE_CORRELATION_SCHEME_AFFINITY:
                    correlation.Scheme = ServiceCorrelationScheme.Affinity;
                    break;
                case NativeTypes.FABRIC_SERVICE_CORRELATION_SCHEME.FABRIC_SERVICE_CORRELATION_SCHEME_ALIGNED_AFFINITY:
                    correlation.Scheme = ServiceCorrelationScheme.AlignedAffinity;
                    break;
                case NativeTypes.FABRIC_SERVICE_CORRELATION_SCHEME.FABRIC_SERVICE_CORRELATION_SCHEME_NONALIGNED_AFFINITY:
                    correlation.Scheme = ServiceCorrelationScheme.NonAlignedAffinity;
                    break;
                default:
                    AppTrace.TraceSource.WriteError("ServiceCorrelationDescription.CreateFromNative", "Invalid service correlation scheme {0}", casted->Scheme);
                    ReleaseAssert.Failfast(string.Format(CultureInfo.CurrentCulture, StringResources.Error_InvalidServiceCorrelationScheme_Formatted, casted->Scheme));
                    break;
            }

            return correlation;
        }

        internal unsafe void ToNative(PinCollection pin, ref NativeTypes.FABRIC_SERVICE_CORRELATION_DESCRIPTION description)
        {
            description.ServiceName = pin.AddObject(this.ServiceName);

            switch (this.Scheme)
            {
                case ServiceCorrelationScheme.Affinity:
                    description.Scheme = NativeTypes.FABRIC_SERVICE_CORRELATION_SCHEME.FABRIC_SERVICE_CORRELATION_SCHEME_AFFINITY;
                    break;
                case ServiceCorrelationScheme.AlignedAffinity:
                    description.Scheme = NativeTypes.FABRIC_SERVICE_CORRELATION_SCHEME.FABRIC_SERVICE_CORRELATION_SCHEME_ALIGNED_AFFINITY;
                    break;
                case ServiceCorrelationScheme.NonAlignedAffinity:
                    description.Scheme = NativeTypes.FABRIC_SERVICE_CORRELATION_SCHEME.FABRIC_SERVICE_CORRELATION_SCHEME_NONALIGNED_AFFINITY;
                    break;
                default:
                    ReleaseAssert.Failfast(string.Format(CultureInfo.CurrentCulture, StringResources.Error_InvalidServiceCorrelationScheme_Formatted, this.Scheme));
                    break;
            }
        }

        /// <summary>
        /// <para> 
        /// Returns a string of the ServiceCorrelationDescription in the form 'ServiceName', 'Scheme'
        /// </para>
        /// </summary>
        /// <returns>
        /// <para>A string representing the ServiceCorrelationDescription object.</para>
        /// </returns>
        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();

            sb.AppendFormat("{0},", this.ServiceName);
            sb.AppendFormat("{0}", this.Scheme);

            return sb.ToString();
        }
    }
}